File: linkedlist/linkedlist.cpp
#include<iostream>
using namespace std;
 class Linkedlist{
    class Node{
        public:
        int data;
        Node* next;
        Node(int data,Node* next=NULL){
            this->data=data;
            this->next=next;
        }
    };
    Node* head;
    
    public:
    Linkedlist(){
        head=nullptr;
        
    }
    void insertathead(int data){
        if(head==nullptr){
            head=new Node(data);
            return;
        }
        Node* newnode= new Node(data,head);
        
        head=newnode;
        
    }
    void insertatlast(int data){
        if(head==nullptr){
            insertathead(data);
            return;
        }
        Node* temp=head;
        while(temp->next!=nullptr){
            temp=temp->next;
        }
        Node* newnode=new Node(data);
        temp->next=newnode;
    }
    void insertatindex(int data,int index){
        if(index==0){
            insertathead(data);
            return;
        }
        Node* temp=head;
        int i=0;
        while(i<index-1){
            temp=temp->next;
            i++;
        }
        Node* newnode=new Node(data);
        newnode->next=temp->next;
        temp->next=newnode;
    }
    void deleteathead(){
        Node* temp=head;
        head=head->next;
        delete temp;
    }
    void deleteatlast(){
        Node* temp=head;
        while(temp->next->next!=nullptr){
            temp=temp->next;
        }
        Node* todelete=temp->next;
        temp->next=nullptr;
        delete todelete;
    }
    void deleteatindex(int index){
        if(index==0){
            deleteathead();
            return;
        }
        Node* temp=head;
        int i=0;
        while(i<index-1){
            temp=temp->next;
            i++;
        }
        Node* todelete=temp->next;
        temp->next=temp->next->next;
        delete todelete;
    }
    int length(){
        Node* temp=head;
        int count=0;
        while(temp!=nullptr){
            count++;
            temp=temp->next;
        }
        return count;
    }
    int get(int index){
        Node* temp=head;
        int i=0;
        while(i<index){
            temp=temp->next;
            i++;
        }
        return temp->data;
    }
    bool search(int key){
        Node* temp=head;
        while(temp!=nullptr){
            if(temp->data==key){
                return true;
            }
            temp=temp->next;
        }
        return false;
    }

    void print(){
        Node* temp=head;
        while(temp!=NULL){
            cout<<temp->data<<" ";
            temp=temp->next;
        }
    }
    Node* getHead() const {
        return head;
    }
 };

 int main(){
     Linkedlist l;
     l.insertathead(1);
     l.insertathead(2);
     l.insertathead(3);
     l.print();
     cout<<endl;
     cout<<l.getHead()->data;
     return 0;
 }

File: linkedlist/linkedlist.java
public class linkedlist {
    class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    Node head;

    public linkedlist() {
        head = null;
    }

    public void insertAtHead(int data) {
        Node newnode = new Node(data);
        newnode.next = head;
        head = newnode;
    }
    public void insertattail(int data){
        Node newnode = new Node(data);
        if (head==null){
           head=new Node(data);
        }
      
        Node temp = head;
        while(temp.next!=null){
            temp=temp.next;
        }
        temp.next=newnode;

    }
    public void insertatindex(int data,int index){
        Node newnode = new Node(data);
        Node temp = head;
        
        while(index-->1){
            temp=temp.next;
        }
        newnode.next=temp.next;
        temp.next=newnode;
    }
    
    public void deleteathead(){
        if(head==null){
            return;
        }
        Node temp = head;
        head = head.next;
        temp = null;
    }
    public void deleteattail(){
        if(head==null){
            return;
        }
        Node temp = head;
        while(temp.next.next!=null){
            temp=temp.next;
        }
        temp.next=null;
    
    }
    public void deleteatindex(int index){
      if(head==null){
          return;
      }
      Node temp =head;
      while(index-->1){
        temp=temp.next;
      }
      Node temp2=temp.next;
      temp.next=temp.next.next;
      temp2=null;

    }
    public int length(){
        Node temp = head;

        int count=0;
        while(temp!=null){
            count++;
            temp=temp.next;
        }
        return count;

    }
    public int getHead(){
        return head.data;
    }
    public int get(int index){
        Node temp = head;
        while(index-->0){
            temp=temp.next;
        }
        return temp.data;
    }
    public boolean search(int key){
        Node temp = head;
        while(temp!=null){
            if(temp.data==key){
                return true;
            }
            temp=temp.next;
        }
        return false;
    }

    public void print() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        linkedlist list = new linkedlist();
        list.insertAtHead(1);
        list.insertAtHead(2);
        list.insertAtHead(3);
        list.insertattail(0);
        list.insertatindex(4,2);
        list.insertattail(7);
        list.insertattail(8);
        list.print();
        list.deleteathead();
        list.print();
        list.deleteattail();
        list.print();
        list.deleteatindex(2);
        list.print();

        System.out.println(list.length());
        System.out.println(list.getHead());
        System.out.println(list.get(2));
        System.out.println(list.search(4));
        System.out.println(list.search(9));
    }
}